File: C/ex3/lrCostFunction.c
#include <stdio.h>
#include <math.h>

double sigmoid(double z);
//static double X[M][N] ;  //array is large then have to use "static"
//static double y[M] ;

double lrCostFunction(int M, int N, double theta[N], double X[M][N], double y[M], double grad[N], double lambda)
{
	int i, j ;
	double   g, JCost=0,  TX, theta2;

	theta2 = 0;
	JCost =0;
	
	for(j=0; j < N; j++)
	        grad[j] = 0;
	
	for(i=0; i < M; i++){
	       TX = 1;
		for(j=0; j < N; j++)
			TX += theta[j]*X[i][j];
			
		g = sigmoid(TX);		
		
		JCost += - y[i]*log(g) - (1.- y[i])* log(1.- g); 
		 
		for(j=0; j < N; j++)
		   grad[j] += X[i][j]*(g - y[i]);		   		   
	}
	
	
	for(j=0; j < N; j++)
		theta2  += pow(theta[j], 2);
		
	JCost = JCost/ M + lambda/2/ M * theta2 ;
	
		
		grad[0] =  grad[0]/ M;
	for(j=0; j < N; j++)
	       grad[j] = grad[j]/ M + lambda/ M * theta[j];	

	
	return JCost;
}

File: Fortran/ex3/lrCostFunction.f90
	subroutine lrCostFunction(theta, y, JCost, grad)
	use mod_common
	implicit none
	
	double precision,dimension(N,1):: theta, grad
	double precision :: JCost
	double precision :: sigmoid, g, TX, theta2
	double precision,dimension(M,1)::y
	integer::i, j
	
	theta2 = 0
	grad = 0
	JCost =0
	
	do i = 1, M
	        TX = 1
		 do j = 1, N
		   TX = TX + X(i,j)*theta(j,1)
		 enddo
		g = sigmoid( TX)	
				 
		JCost = JCost - y(i,1) * log(g)- (1- y(i,1))* log(1- g)
		
		do j = 1, N
		   grad(j,1) = grad(j,1) + X(i,j)* (g - y(i,1))
		enddo
	enddo	
	
		do j = 1, N
		   theta2 = theta2 + theta(j,1)**2
		 enddo
		 
	  JCost = JCost/m + lambda/2/m *theta2
	  
	           grad = grad/m
		do j = 1, N
		      grad(j,1) = grad(j,1) + lambda /m *theta(j,1)
		enddo
		
	end
